src/zookeeperLock.ts

```diff
@@ -467,66 +635,67 @@ export class ZookeeperLock extends EventEmitter {
-                        .map((l) => {
-                            return ZookeeperLock.getSequenceNumber(l);
-                        })
-                        .filter((l) => {
-                            return l >= 0;
-                        });
-
-                    debuglog(`${path}/${this.key}: lock sequence: ${JSON.stringify(sequence)}`);
-
-                    const mySeq = ZookeeperLock.getSequenceNumber(this.key);
-
-                    // todo: fix me for max concurrent holders...
-                    // make sure we are first...
-                    const min = sequence.reduce((acc, elem) => {
-                        return Math.min(acc, elem);
-                    }, mySeq);
-
-                    debuglog(`${path}/${this.key}: checking ${mySeq} less than ${min} + ${this.config.maxConcurrentHolders}`);
-                    if (mySeq < (min + this.config.maxConcurrentHolders)) {
-                        debuglog(`${path}/${this.key}: ${mySeq} can grab the lock on ${path}`);
-                        return resolve(true);
-                    } else if (this.config.failImmediate) {
-                        this.locked = false;
-                        this.destroyed = true;
-                        debuglog(`${path}/${this.key}: failing immediately`);
-                        return reject(new ZookeeperLockAlreadyLockedError('already locked', path));
+                            const sequence = this.filterLocks(locks)
+                                .map((l) => {
+                                    return ZookeeperLock.getSequenceNumber(l);
+                                })
+                                .filter((l) => {
+                                    return l >= 0;
+                                });
+
+                            debuglog(`${path}/${this.key}: lock sequence: ${JSON.stringify(sequence)}`);
+
+                            const mySeq = ZookeeperLock.getSequenceNumber(this.key);
+
+                            const sorted : Array<number> = sequence.sort((a, b) => {
+                                return a - b;
+                            });
+                            const offset = Math.min(sorted.length, this.config.maxConcurrentHolders);
+                            const min = sorted[offset - 1];
+
+                            debuglog(`${path}/${this.key}: checking ${mySeq} less than ${min} + ${this.config.maxConcurrentHolders}`);
+                            if (mySeq <= min) {
+                                debuglog(`${path}/${this.key}: ${mySeq} can grab the lock on ${path}`);
+                                return resolve(true);
+                            } else if (this.config.failImmediate) {
+                                this.changeState(ZookeeperLock.States.ALREADY_LOCKED);
+                                debuglog(`${path}/${this.key}: failing immediately`);
+                                this.unlock().finally(() => {
```

Comment:
not exactly what was leaking, but no i wasn't doing it before. I did this to make the call to remove the entry from the lock path immediately, since it doesn't need to stay in line waiting to be the lowest sequence in `failImmediate` mode. Probably more proper would be to call `this.client.remove` like unlock is doing, but without any of the other steps, but it does make the `failImmediate` mode a bit more braindead to use like this, as it cleans up after itself, no call to `lock.unlock` or `lock.destroy` necessary I suppose..